### PR TITLE
Add cibuildwheel workflow

### PR DESCRIPTION
**Pull request by sevagh**
This re-adds the cibuildwheel workflow (and related changes to setup.py) from PR #27 

PR #27 was merged prematurely - I missed the discussion around patching `__version__`. This new PR has a fix incorporated so that `__version__` matches the same version metadata encoded by `setup(...)`.

### CODE REVIEW

**Comment by sevagh** [comment]
The `__version__` adjustment works! @wence- (cc @jakirkham @gmarkall @vyasr ), this resolves the conversation from the last PR: https://github.com/rapidsai/ptxcompiler/pull/27#discussion_r984378278 

```
(ptxcompiler-cp38-test) sevagh@pop-os:~/Downloads/ptxcwheel$ pip install ./ptxcompiler_cu11-0.7.0.1668704666-cp38-cp38-manylinux_2_17_x86_64.manylinux2014_x86_64.whl
Processing ./ptxcompiler_cu11-0.7.0.1668704666-cp38-cp38-manylinux_2_17_x86_64.manylinux2014_x86_64.whl
Installing collected packages: ptxcompiler-cu11
Successfully installed ptxcompiler-cu11-0.7.0.1668704666
WARNING: You are using pip version 22.0.4; however, version 22.3.1 is available.
You should consider upgrading via the '/home/sevagh/venvs/ptxcompiler-cp38-test/bin/python -m pip install --upgrade pip' command.
(ptxcompiler-cp38-test) sevagh@pop-os:~/Downloads/ptxcwheel$
(ptxcompiler-cp38-test) sevagh@pop-os:~/Downloads/ptxcwheel$
(ptxcompiler-cp38-test) sevagh@pop-os:~/Downloads/ptxcwheel$
(ptxcompiler-cp38-test) sevagh@pop-os:~/Downloads/ptxcwheel$
(ptxcompiler-cp38-test) sevagh@pop-os:~/Downloads/ptxcwheel$ python -c 'import ptxcompiler; print(ptxcompiler.__version__)
'
0.7.0.1668704666
```

File: setup.py [hunk 1]
```diff
@@ -48,11 +48,27 @@
...
+    # borrow a similar hack from dask-cuda: https://github.com/rapidsai/dask-cuda/blob/b3ed9029a1ad02a61eb7fbd899a5a6826bb5cfac/setup.py#L12-L31
+    orig_get_versions = versioneer.get_versions
+
+    version_override = os.environ.get("RAPIDS_PY_WHEEL_VERSIONEER_OVERRIDE", "")
+
+    def get_versions():
+        data = orig_get_versions()
+        if version_override != "":
+            data["version"] = version_override
+        return data
```
Comment (wence-) [inline setup.py:61]: I think that we want to complain more loudly if something is in the environment but empty. So how about writing this like so:

```python
if "RAPIDS_PY_WHEEL_VERSIONEER_OVERRIDE" in os.environ:
    orig_get_versions = versioneer.get_version

   version_override = os.environ["RAPIDS_PY_WHEEL_VERSIONEER_OVERRIDE"]
   if version_override == "":
       raise RuntimeError("RAPIDS_PY_WHEEL_VERSIONEER_OVERRIDE in environment but empty")
   
   def get_versions():
       data = orig_get_versions():
       data["version"] = version_override
       return data

    versioneer.get_versions = get_versions
```

Comment (wence-) [reply setup.py:61]: Or is it that we only need this for nightlies and for actual tagged version releases we don't override things?

**Review by wence- — CHANGES_REQUESTED** [review]
Possibly changes required, unless I misunderstand how some of this fits together.

File: .github/workflows/wheels.yml [hunk 2]
```diff
@@ -0,0 +1,52 @@
+name: ptxcompiler wheels
+
+on:
+  workflow_call:
+    inputs:
+      versioneer-override:
+        type: string
+        default: ''
```
[the remark this answers is not in the document]
Comment (sevagh) [reply .github/workflows/wheels.yml:8]: They are set by the nightly pipeline that builds all pip wheels: https://github.com/rapidsai/actions/blob/main/.github/workflows/pip-wheels-pipeline.yml

Every pip wheel workflow is basically (at least, during the 22.12 release), _only_ intended to be invoked by external repositories.

If the input is unspecified, such that it is the empty string, then this is a 'nightly build'; the env var `RAPIDS_PY_WHEEL_VERSIONEER_OVERRIDE` will be assembled from `git describe --tag --abbrev=0` (dropping alphabet letters e.g. `v0.7.00a -> 0.7.0` + epoch timestamp of the nightly build

If the input is non-empty, it is used directly as `RAPIDS_PY_WHEEL_VERSIONEER_OVERRIDE`

https://github.com/rapidsai/shared-action-workflows/blob/main/cibuildwheel/action.yml#L90
https://github.com/rapidsai/shared-action-workflows/blob/main/cibuildwheel/action.yml#L97 

Basically, there is no circumstance where we let versioneer do its own thing, at this point in time - but that's just to save time and succeed at the 22.12 release by continuing the hacks/bandaids of the 22.10 release.

For future releases (23.02+), we can (and definitely should) address all of these pain points such that we use our tools appropriately.

File: setup.py (hunk 1 above)
Comment (sevagh) [reply setup.py:61]: It's a good suggestion, I will implement it. As I mentioned above, right now we always override (as a bandaid), so it's better to be safe about it.

**Review by bdice — APPROVED** [review]

File: .github/workflows/wheels.yml (hunk 2 above)
[the remark this answers is not in the document]
Comment (sevagh) [reply .github/workflows/wheels.yml:8]: Let me know if the explanation is sufficient @wence-

[the remark this answers is not in the document]
Comment (wence-) [reply .github/workflows/wheels.yml:8]: Thanks, makes sense.

**Comment by jakirkham** [comment]
Think we still need a very basic `pyproject.toml` here

**Comment by vyasr** [comment]
> Think we still need a very basic `pyproject.toml` here

@jakirkham could you explain why you think it is necessary in this PR? We definitely want to update the build system eventually, but I don't think it is necessary for the wheel building to work. With no pyproject.toml the build just falls back to the legacy setuptools-based builds, which will still produce valid wheels.

**Comment by jakirkham** [comment]
For clarity am just meaning something like this

```TOML
[build-system]
requires = ["setuptools", "wheel"]
build-backend = "setuptools.build_meta
```

**Comment by vyasr** [comment]
> For clarity am just meaning something like this
> 
> ```toml
> [build-system]
> requires = ["setuptools", "wheel"]
> build-backend = "setuptools.build_meta
> ```

Yup, and I'm happy to make that change to this repo soon, but I'm not sure why it's in scope for this PR. This project is still using a fully legacy setuptools-based build since it has a setup.py file and no pyproject.toml, and all of our tooling should still support that, including `pip wheel` (just with deprecation warnings). If it isn't necessary to get wheel building to work I think it's out of scope for this PR and can be done in a follow-up. Do you see a reason why the current state of the code might fail to build correctly without the file?

**Comment by jakirkham** [comment]
If we'd prefer to use the legacy backend, that seems fine

> ```toml
> [build-system]
> requires = ["setuptools", "wheel"]
> build-backend = "setuptools.build_meta:__legacy__"
> ```

The build tooling in the Python ecosystem is already moving in this direction. It would be better to get out ahead of it than get bitten later by something changing on us

Also being explicit about what build backend we are using will help communicate this to others as well as to keep track ourselves. Especially when doing the work to migrate to the newer (preferred) build backend

File: setup.py [hunk 3]
```diff
@@ -45,14 +45,32 @@
...
+        raise RuntimeError(
+            "RAPIDS_PY_WHEEL_VERSIONEER_OVERRIDE in environment but empty"
+        )
+
+    def get_versions():
+        data = orig_get_versions()
+        data["version"] = version_override
+        return data
+
+    versioneer.get_versions = get_versions
```
Comment (wence-) [inline setup.py:65]: Does the shared actions workflow still set up these env vars? I know that many other packages have been migrated off versioneer.

**Review by wence- — APPROVED** [review]
I think this looks ok, sorry for leaving it lying for so long.

File: setup.py (hunk 3 above)
Comment (vyasr) [reply setup.py:65]: Good question. The short answer is no, those env vars aren't set up, but it doesn't matter here.

The longer answer: the other packages building wheels have been migrated away from versioneer and to a large extent away from setup.py as well. Almost all of their packaging metadata is in pyproject.toml now. The workflows modify that file directly since it's a well-defined config file instead of a code file. If we wanted ptxcompiler to go the same route, then we would need to update this package in that way as well. However, we don't use the shared workflows to release ptxcompiler. Since the release cycle is slower and more irregular than the rest of RAPIDS, and since the build is much simpler, we create wheels on-demand using some scripts that can be run locally. I've ensure that those scripts will still work with this versioning approach, so that should be sufficient for everything to "just work" for us.

Comment (wence-) [reply setup.py:65]: Thanks!
